### ISSUE TITLE
[Cloud Block Service] AWS Volume shows status available. On modifying throws error cannot be modified when state is MODIFYING or OPTIMIZING.

### ISSUE THREAD

**Issue by anvithks** [sodafoundation/multi-cloud#1120]
**Issue/Feature Description:**
When a volume that is newly created is modified the status is shown as available. 
After the status is shown available if another modify request is sent the backend throws an error which has no information (panic error). On checking the docker logs the below error is shown.

On trying to modify a volume which is in MODIFYING or OPTIMIZING state the API throws an error.

These states can be fetched from the cloud backend and returned in the API to give the user correct information

```
[2020-09-28T21:49:21+05:30] [ERROR] [/root/gopath/src/github.com/sodafoundation/multi-cloud/contrib/datastore/block/aws/aws.go] [aws.(*AwsAdapter).UpdateVolume():318] [PID:1] Error in updating volume: {
  Iops: 0,
  Size: 600,
  VolumeId: "vol-08986585387cd1eff",
  VolumeType: "st1"
}%!(EXTRA *awserr.requestError=IncorrectModificationState: Volume vol-08986585387cd1eff cannot be modified in modification state MODIFYING
	status code: 400, request id: 7e2390e5-7d2c-415b-af3a-9bd49857b5bd)
[2020-09-28T21:49:21+05:30] [ERROR] [/root/gopath/src/github.com/sodafoundation/multi-cloud/contrib/datastore/block/aws/aws.go] [aws.(*AwsAdapter).UpdateVolume():322] [PID:1] IncorrectModificationState: Volume vol-08986585387cd1eff cannot be modified in modification state MODIFYING
```

**Why this issue to fixed / feature is needed(give scenarios or use cases):**
If the state is shown as available then the user will try to modify. we are already showing creating and updating as intermediate states wherein a user will expect some change. Similarly showing the state as modifying or optimizing as per the cloud vendor will give more accurate information to the user and will also help in preventing making API calls to the backend when certain states do not allow it.

**How to reproduce, in case of a bug:**

**Other Notes /  Environment Information: (Please give the env information, log link or any useful information for this issue)**

**Comment by kumarashit** [comment]
Should be fixed in next release.
**Reason for not fixing in Greenland release** Less chances that so frequent modification will be done in actual/production environment

**Comment by himanshuvar** [comment]
@anvithks @kumarashit This is as expected from AWS Cloud Backend

@anvithks Could you please update the issue with the Response from API

**Comment by anvithks** [comment]
The issue is wrt the status in the SODA API response not matching the AWS Cloud backend error.
The SODA API returns status as available. When the user tries to modify it throws above error.

You can check the scenario in the below screenshots.

1. Created a GP volume with 100GB : available
![image](https://user-images.githubusercontent.com/19162717/94515831-1ba7a200-0242-11eb-9c62-64ad52192763.png)

2. Docker logs
![image](https://user-images.githubusercontent.com/19162717/94515887-3bd76100-0242-11eb-9b57-a1e0fae08e10.png)

3. API response in the Modify volume page
![image](https://user-images.githubusercontent.com/19162717/94515944-58739900-0242-11eb-8556-63f116b0e5b8.png)


4. Docker logs in the modify volume page.
![image](https://user-images.githubusercontent.com/19162717/94515972-69240f00-0242-11eb-81e6-f4ecff3172e7.png)


5. Modified the size to 500GB. Status is updating
![image](https://user-images.githubusercontent.com/19162717/94516000-7a6d1b80-0242-11eb-8e54-084254c10f48.png)

6. API response is updating.
![image](https://user-images.githubusercontent.com/19162717/94516024-8953ce00-0242-11eb-89ce-0e321345acdf.png)

7. Status is now available.
![image](https://user-images.githubusercontent.com/19162717/94516067-a12b5200-0242-11eb-9581-22bc1f7c9dc5.png)

Docker logs
![image](https://user-images.githubusercontent.com/19162717/94516122-bc965d00-0242-11eb-9a5e-b6d80c27e7dc.png)

8. Modify request sent. Size increased to 750GB. 
Status is available but the cannot modify as modification state is OPTIMIZING.
![Uploading image.png…]()

**Comment by himanshuvar** [comment]
@anvithks Could you please share Response for Step 8

**Comment by anvithks** [comment]
The API response
![image](https://user-images.githubusercontent.com/19162717/94516562-b81e7400-0243-11eb-9d29-820cc9439b7e.png)

**Comment by himanshuvar** [comment]
@anvithks Is this PUT Response for 750 GB?, Context Deadline exceeded was fixed using go routine. PUT API should return some response in terms of Error.

**Comment by himanshuvar** [comment]
@anvithks I suggest try same operations on AWS cloud Backend, and notice the behavior of Request/Response from AWS console.

**Comment by anvithks** [comment]
Yes @himanshuvar. This is for PUT. This setup was made last night.

**Comment by anvithks** [comment]
> @anvithks I suggest try same operations on AWS cloud Backend, and notice the behavior of Request/Response from AWS console.

SODA clients will communicate with the SODA backend. The response we receive from the SODA backend willl be considered the same as that of the cloud vendor.

**Comment by himanshuvar** [comment]
> 
> 
> > @anvithks I suggest try same operations on AWS cloud Backend, and notice the behavior of Request/Response from AWS console.
> 
> SODA clients will communicate with the SODA backend. The response we receive from the SODA backend willl be considered the same as that of the cloud vendor.

Yes, I understand that SODA APIs & Clients communication should be in sync. If SODA clients expect something on some operations that is not given back from SODA APIs is certainly an issue in the APIs
I have suggested the Cloud operation just to make our understanding clear about the AWS cloud operations. What is the expected behavior for their call operations. i.e.
_On modifying throws error cannot be modified when state is MODIFYING or OPTIMIZING_ -> How AWS response to this operation.

**Comment by anvithks** [comment]
This is how AWS console works. Hope this helps.

Modify prompt
![image](https://user-images.githubusercontent.com/19162717/94538596-d2674a80-0261-11eb-9f53-898728c183c6.png)

Modify Success
![image](https://user-images.githubusercontent.com/19162717/94538632-db581c00-0261-11eb-8e2c-57b2d87e103b.png)

**Intermediate state. Volume is available but there is a modification state maintained and returned in the API. Denoted with yellow color next to available in the list.**
![image](https://user-images.githubusercontent.com/19162717/94538699-f62a9080-0261-11eb-91f0-84606c227e7f.png)

**API  with modification state**
![image](https://user-images.githubusercontent.com/19162717/94538738-03477f80-0262-11eb-8afa-5611d917a41f.png)

Error when trying to modify the volume when in optimizing state.
![image](https://user-images.githubusercontent.com/19162717/94538784-122e3200-0262-11eb-8f91-a676af2fa795.png)

Error from API
![image](https://user-images.githubusercontent.com/19162717/94538863-2bcf7980-0262-11eb-8707-5f1290f818cb.png)

**Comment by himanshuvar** [comment]
Thanks @anvithks for confirming the operations from AWS Cloud Backend.
I have tested the exact operations with master branch using POSTMAN client, Please find the results below.
Hope you are also executing the above results on latest code.

1. Create a 100GB gp2 Volume using SODA multi-cloud Block API
<img width="792" alt="1" src="https://user-images.githubusercontent.com/8712951/94544084-7227d700-0268-11eb-9837-925c47019cd8.PNG">

 docker logs for creating a 100GB Volume
<img width="864" alt="2" src="https://user-images.githubusercontent.com/8712951/94544192-91266900-0268-11eb-84f5-4b5dbae8e0c8.PNG">

Volume is created at AWS backend, can be seen in AWS console as below
<img width="795" alt="3" src="https://user-images.githubusercontent.com/8712951/94544358-c3d06180-0268-11eb-84ce-f960c3ac897e.PNG">

2. Get Volume, created in Step 1 using SODA multi-cloud Block API
<img width="786" alt="4" src="https://user-images.githubusercontent.com/8712951/94544464-e498b700-0268-11eb-9dc4-7210b36a958f.PNG">

3. Update Volume to 500GB using SODA multi-cloud Block API
<img width="791" alt="5" src="https://user-images.githubusercontent.com/8712951/94544698-2de90680-0269-11eb-8a8e-e7dde8ea7b2a.PNG">

docker logs
<img width="857" alt="6" src="https://user-images.githubusercontent.com/8712951/94544839-6092ff00-0269-11eb-84f5-cb17ab70fe71.PNG">

Volume is Updated at AWS backend, can be seen in AWS console as below
<img width="795" alt="7" src="https://user-images.githubusercontent.com/8712951/94544884-71dc0b80-0269-11eb-8e40-682eb1b78742.PNG">

4. Get Volume, updated in Step 3 using SODA multi-cloud Block API
<img width="794" alt="8" src="https://user-images.githubusercontent.com/8712951/94545066-abad1200-0269-11eb-8edb-1bc1e1be2548.PNG">

5. Update Volume to 750GB using SODA multi-cloud Block API. **THIS IS A FAILED OPERATION**
<img width="784" alt="9" src="https://user-images.githubusercontent.com/8712951/94545259-f038ad80-0269-11eb-940a-93fecef6926b.PNG">

This Calls returns the error with the valid reasons response from backend 

docker logs
<img width="858" alt="10" src="https://user-images.githubusercontent.com/8712951/94545580-58878f00-026a-11eb-97bd-37ba05f2ac4b.PNG">

6. Get Volume using SODA multi-cloud Block API, Volume is in its old STATE which is an expected behavior 
<img width="790" alt="11" src="https://user-images.githubusercontent.com/8712951/94545646-70f7a980-026a-11eb-96f3-a07fbf00e504.PNG">

Volume is in its old STATE which is an expected behavior at AWS backend, can be seen in AWS console as below
<img width="794" alt="12" src="https://user-images.githubusercontent.com/8712951/94545878-c03dda00-026a-11eb-8e53-e04b9bd6a5f3.PNG">
Volume Resource is maintained at the SODA Backend as AWS Backend

@kumarashit Please look into above findings. Please let me know if above SODA multi-cloud API's request/response are correct or not. Please let us know if any other API behavior expectated for these API calls.

@kumarashit @anvithks Let me know if there is any gap from my side in understanding the expected SODA multi-cloud backend APIs requirements.

**Comment by anvithks** [comment]
AWS adds the modification state to the response when the volume is modifying as shown here.
![image](https://user-images.githubusercontent.com/19162717/94538738-03477f80-0262-11eb-8afa-5611d917a41f.png)

if the modification state is returned in SODA API then we can preempt making any requests for modification when the volume is in any of the intermediate states. 
AWS uses the colored dot next to the state name to indicate the intermediate state. An available volume is shown in green.
An available volume in OPTIMIZING or MODIFYING state is shown with a yellow dot. 

If the client receives the modification state object as shown above then making another request for modification can be prevented.

**Comment by himanshuvar** [comment]
@anvithks This is another approach to leverage the STATE model from a cloud vendor into SODA, but I don't think this Approach is feasible in SODA because of the following reasons:
1. SODA doesn't support Operation Handling. Its just a placeholder for the provisioning & metadata etc.
2. Approach is not scalable for extending the other Cloud vendor support as the other cloud vendor might not support the same STATE models as AWS like OPTIMIZING or MODIFYING. So we can't go with AWS biasing in SODA unless that is standard across multi-cloud vendors.
3. The PUT API calls can't be restricted based on the STATE model only. There are other factors/Errors need to be considered like Operation permitted within time limit that anyways are handled from cloud backend so as by SODA multi-cloud API. 

Instead of confusing the Users for different states, the API request/response can be used for managing API calls that is already supported in the API. Client should utilize the API response properly.

**Comment by anvithks** [comment]
@himanshuvar 
>SODA doesn't support Operation Handling. Its just a placeholder for the provisioning & metadata etc.

Could you please elaborate a bit on this. Not sure I understand this correctly.

>Approach is not scalable for extending the other Cloud vendor support as the other cloud vendor might not support the same STATE models as AWS like OPTIMIZING or MODIFYING. So we can't go with AWS biasing in SODA unless that is standard across multi-cloud vendors.

True. But we need to have a mechanism that would allow cloud vendor specific restrictions, validations or responses to be available to the client. For ex: in the current volume modify issue, a user who may be using SODA APIs will not be able to know that the status is available but in another intermediate state. Even if it is a valid scenario to throw an error it doesn't add up to a clean experience. 

>The PUT API calls can't be restricted based on the STATE model only. There are other factors/Errors need to be considered like Operation permitted within time limit that anyways are handled from cloud backend so as by SODA multi-cloud API.

With a front end client you do, in fact, get some amount of logic that can control the number of calls that need to be made to the backend. using the APIs directly using any REST client will still work as you explained above. If some of the limits and validations are known to us then we can leverage the same in the SODA APIs to make the usage more inline with e cloud backend. If the SODA APIs will only pass the error from the cloud vendor backend then this may lead to many wasted calls from the client.

> Instead of confusing the Users for different states,  the API request/response can be used for managing API calls that is already supported in the API. Client should utilize the API response properly.  

The users will in fact be in  a confused state when they see an error for an available volume. This is my main concern. if the status is available and there is no way for the user to know that the volume may not be modified how do we inform the user abou the same.
